### PR TITLE
PCHR-2209:  Fix Issues with updating An Already Approved TOIL

### PR DESCRIPTION
**Pull request by tunbola**
## Overview
If you accrue TOIL as a staff and your manager approves it and saves it. Then maybe the manager thinks he made a mistake and changes the status back to 'Information Requested' and tries updating the TOIL request or even tries to update the value of the TOIL accrued, An error appears that 'The TOIL amount plus all approved TOIL for current period is greater than the maximum for this Absence Type'.

## Before
To illustrate the Bug, consider the following table showing TOIL accrued for a contact in a given period for a particular Absence Type with max_leave_accrual of 9.

| TOIL Name  | Status | Amount Accrued |
| ------------- | ------------- | ------------- |
| TOIL A  | Approved  |4|
| TOIL B  | Approved  |3|

To accrue another TOIL called TOIL C with amount 3, the validation goes like this:

**Total TOIL Already Approved for Absence Type**  = 4 (TOIL A) + 3 (TOIL B) = 7;
**Total projected TOIL** = 7(Total Approved) + 3(About to be Accrued, TOIL C) = 10
if the Total projected TOIL > max_leave_accrual for Absence Type, An exception will be thrown, in this case 10 > 9.

Currently if you try to update TOIL B say to 4
Total TOIL Already Approved for Absence Type  = 4 (TOIL A) + 3 (TOIL B) = 7;
Total projected TOIL = 7(Total Approved) + 4(About to be Accrued, TOIL B) = 11
TOIL B is treated as another TOIL to be accrued and it is counted as approved TOIL and also in projected TOIL, hence exception is thrown.

## After
When updating an Already approved TOIL, the amount of the TOIL previously accrued is deducted from the approved TOIL before the new amount is added to the projected TOIL.

To illustrate with a continuation of the example above:

Currently if you try to update TOIL B say to 4
Total TOIL Already Approved for Absence Type  = 4 (TOIL A) + 3 (TOIL B) = 7;
Total projected TOIL = 7(Total Approved) + 4(About to be Accrued, TOIL B) - 3(Previous value for TOIL B)= 8
Since 8 is not greater than max_leave_accrual of 10, the update goes through.

- [X] Tests Pass

### CODE REVIEW

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/LeaveRequest.php [hunk 1]
```diff
@@ -282,6 +287,15 @@ private static function validateTOILToAccruedAmountIsValid($params) {
     );
     $totalProjectedToilForPeriod = $totalApprovedToilForPeriod + $params['toil_to_accrue'];
 
+    if ($oldToilRequest && $oldToilRequest->id && self::isAlreadyApproved($oldToilRequest)) {
```
Comment (mickadoo) [inline uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/LeaveRequest.php]: This is a very nice block of code, clear and concise :-)

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/LeaveRequest.php [hunk 2]
```diff
@@ -1040,4 +1054,45 @@ public function addSelectWhereClause() {
...
+  private static function getApprovalStatuses() {
+    $leaveStatuses = array_flip(self::buildOptions('status_id', 'validate'));
+
+    return [$leaveStatuses['approved'], $leaveStatuses['admin_approved']];
+  }
+
+  /**
+   * Checks whether a leave request has been approved.
+   *
+   * @param \CRM_HRLeaveAndAbsences_BAO_LeaveRequest $leaveRequest
```
Comment (mickadoo) [inline uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/LeaveRequest.php]: CRM_HRLeaveAndAbsences_BAO_LeaveRequest might be a candidate for an alias

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/LeaveRequest.php [hunk 3]
```diff
@@ -1040,4 +1054,45 @@ public function addSelectWhereClause() {
...
+   *
+   * @return bool
+   */
+  private static function isAlreadyApproved(LeaveRequest $leaveRequest) {
```
Comment (mickadoo) [inline uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/LeaveRequest.php]: Nice method

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php [hunk 4]
```diff
@@ -2335,4 +2335,292 @@ public function testToilToAccrueAmountIsSavedCorrectlyWhenAmountToAccrueIsAFloat
     sort($toilAmounts);
     $this->assertEquals($toilAmounts, $toilToAccrueAmounts);
   }
+
+  /**
+   * @expectedException CRM_HRLeaveAndAbsences_Exception_InvalidLeaveRequestException
+   * @expectedExceptionMessage The TOIL amount plus all approved TOIL for current period is greater than the maximum for this Absence Type
+   */
+  public function testUpdatingToilThrowsExceptionWhenUpdatingApprovedToilWithAToilAmountGreaterThanWhatWasInitiallyApprovedAndTotalToilToBeAccruedIsGreaterThanMaximumForTheAbsenceType() {
```
Comment (mickadoo) [inline uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php]: Woah, first prize for longest method name. I know you mentioned before about long method name, but is there any limit? :grin:

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php [hunk 5]
```diff
@@ -2335,4 +2335,292 @@ public function testToilToAccrueAmountIsSavedCorrectlyWhenAmountToAccrueIsAFloat
...
+      'end_date' => CRM_Utils_Date::processDate('+15 days'),
+    ]);
+
+    $absenceType = AbsenceTypeFabricator::fabricate([
+      'title' => 'Type 1',
+      'allow_accruals_request' => true,
+      'max_leave_accrual' => 4
+    ]);
+
+    $periodEntitlement = LeavePeriodEntitlementFabricator::fabricate([
```
Comment (mickadoo) [inline uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php]: The fabricators look nice

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php [hunk 6]
```diff
@@ -2335,4 +2335,292 @@ public function testToilToAccrueAmountIsSavedCorrectlyWhenAmountToAccrueIsAFloat
...
+
+    $toilRequest1 = LeaveRequestFabricator::fabricateWithoutValidation($params, true);
+    $params['from_date'] = CRM_Utils_Date::processDate('friday');
+    $params['to_date'] = CRM_Utils_Date::processDate('friday');
+    $toilRequest2 = LeaveRequestFabricator::fabricateWithoutValidation($params, true);
+
+    //two TOIL requests created and total accrued TOIL is 2 + 2 = 4
+    $this->assertEquals(4, $periodEntitlement->getBalance());
+
+    //User tries to update the second TOIL with a toil_to_accrue value of 3. Initial value for this TOIL is 2
```
Comment (mickadoo) [inline uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php]: For long comments like this maybe the `/**` style is better

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php (hunk 4 above)
Comment (omarabuhussein) [inline uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php]: I agree this is insane @tunbola  😄

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php [hunk 7]
```diff
@@ -2335,4 +2335,292 @@ public function testToilToAccrueAmountIsSavedCorrectlyWhenAmountToAccrueIsAFloat
...
+    //period 1 during the validation and then most likely the TOIL would be accrued for Period 1.
+    //
+    //Exception is thrown because Max leave accrual is 5 and total TOIL already accrued for period 1 is 4
+    //So the code is seeing this TOIL as a new TOIL to be accrued for this period not as an Approved TOIL to be updated which is correct.
+    //4(already accrued) + 2(to be accrued)  = 6 and is greater than max leave accrual for Period 1.
+    $params['id'] = $toilRequest2Period2->id;
+    $params['toil_to_accrue'] = 2;
+    $params['from_date'] = CRM_Utils_Date::processDate('2016-01-06');
+    $params['to_date'] = CRM_Utils_Date::processDate('2016-01-06');
+    LeaveRequestFabricator::fabricate($params, true);
```
Comment (mickadoo) [inline uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php]: Nice coverage of lots of cases

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php (hunk 4 above)
Comment (tunbola) [inline uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php]: Oh boy. We have a lot of that kinda test names in L&A. Descriptive at a glance :)

Comment (omarabuhussein) [inline uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php]: I would say destructive 😄

Comment (tunbola) [inline uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php]: BTW using `phpuint4 --testdox` makes long names useful as you can use that to view all tests in a class in readable English.

Comment (omarabuhussein) [inline uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php]: the thing is with an open source project like this , not everyone going to use phpunit4 --testdox to read through the tests .. and even so , I think we should code for people not for tools ,  so nothing justify writing a method such long even for tests and even if we are assumed everyone is going to use --testdox.

Comment (tunbola) [inline uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php]: Well Davi probably needs to be brought into the discussion then. Like I said earlier, Its favoured in L&A.

**Pull request merged**